### PR TITLE
Add more property descriptions to resources

### PR DESCRIPTION
**Pull request by tas50**
Slowly we'll get them all.

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: lib/chef/resource/ifconfig.rb [hunk 1]
```diff
@@ -35,23 +35,63 @@ class Ifconfig < Chef::Resource
...
+
+      property :hwaddr, String,
+               description: "The hardware address for the network interface."
+
+      property :mask, String,
+               description: "The decimal representation of the network mask. For example: 255.255.255.0."
+
+      property :family, String,
+               default: "inet", introduced: "14.0",
+               description: "Networking family option for Debian-based systems; for example: inet or inet6."
```
Comment (coderanger) [inline lib/chef/resource/ifconfig.rb]: Can I be super pedantic that you used two different styles of `for example` here and the one above? O:-)

File: lib/chef/resource/ifconfig.rb [hunk 2]
```diff
@@ -35,23 +35,63 @@ class Ifconfig < Chef::Resource
...
+               description: "Bring up the network interface on boot."
+
+      property :network, String,
+               description: "The address for the network interface."
+
+      property :bootproto, String,
+               description: "The boot protocol used by a network interface."
+
+      property :onparent, String,
+               description: "Bring up the network interface when its parent interface is brought up."
```
Comment (coderanger) [inline lib/chef/resource/ifconfig.rb]: Since this isn't a boolean, what kind of value would go in here?

File: lib/chef/resource/launchd.rb [hunk 3]
```diff
@@ -30,16 +30,39 @@ class Launchd < Chef::Resource
...
-      property :mode, [String, Integer]
-      property :owner, [String, Integer]
-      property :path, String
-      property :source, String
-      property :session_type, String
+      property :label, String,
+               identity: true, name_property: true,
+               description: "The unique identifier for the job."
+
+      property :backup, [Integer, FalseClass],
```
Comment (coderanger) [inline lib/chef/resource/launchd.rb]: This probably needs a default value.

File: lib/chef/resource/launchd.rb [hunk 4]
```diff
@@ -30,16 +30,39 @@ class Launchd < Chef::Resource
...
+
+      property :group, [String, Integer],
+               description: "When launchd is run as the root user, the group to run the job as. If the username property is specified and this property is not, this value is set to the default group for the user."
+
+      property :plist_hash, Hash,
+               introduced: "12.19",
+               description: "A Hash of key value pairs used to create the launchd property list."
+
+      property :mode, [String, Integer],
+               description: "A quoted 3-5 character string that defines the octal mode. For example: '755', '0755', or 00755."
```
Comment (coderanger) [inline lib/chef/resource/launchd.rb]: It appears this takes a integer too, and `quoted` is odd (makes it sound like you need to do `"'755'"` or something.

File: lib/chef/resource/launchd.rb [hunk 5]
```diff
@@ -50,96 +73,180 @@ class Launchd < Chef::Resource
...
-        types = %w{daemon agent}
-
-        unless types.include?(type)
-          error_msg = "type must be daemon or agent"
-          raise Chef::Exceptions::ValidationFailed, error_msg
-        end
-        type
-      }
+      property :start_calendar_interval, [Hash, Array],
+               description: "A Hash (similar to crontab) that defines the calendar frequency at which a job is started.",
```
Comment (coderanger) [inline lib/chef/resource/launchd.rb]: Can also take an array.

File: lib/chef/resource/launchd.rb [hunk 6]
```diff
@@ -50,96 +73,180 @@ class Launchd < Chef::Resource
...
+               description: "An array of non-empty directories which, if any are modified, will cause a job to be started."
+
+      property :root_directory, String,
+               description: "chroot to this directory, and then run the job."
+
+      property :run_at_load, [ TrueClass, FalseClass ],
+               description: "Launch a job once (at the time it is loaded)."
+
+      property :sockets, Hash,
+               description: "A Hash of on-demand sockets that notify launchd when a job should be run."
```
Comment (coderanger) [inline lib/chef/resource/launchd.rb]: If we have one, an example of the format would be good.

File: lib/chef/resource/launchd.rb [hunk 7]
```diff
@@ -50,96 +73,180 @@ class Launchd < Chef::Resource
...
+               description: "Start a job every time a file system is mounted."
+
+      property :throttle_interval, Integer,
+               description: "The frequency (in seconds) at which jobs are allowed to spawn."
+
+      property :time_out, Integer,
+               description: "The amount of time (in seconds) a job may be idle before it times out. If no value is specified, the default timeout value for launchd will be used."
+
+      property :umask, Integer,
+               description: "A decimal value to pass to umask before running a job."
```
Comment (coderanger) [inline lib/chef/resource/launchd.rb]: Doesn't need to be specifically decimal I would think.

File: lib/chef/resource/launchd.rb [hunk 8]
```diff
@@ -50,96 +73,180 @@ class Launchd < Chef::Resource
...
+               description: "When launchd is run as the root user, the user to run the job as."
+
+      property :wait_for_debugger, [ TrueClass, FalseClass ],
+               description: "Specify if launchd has a job wait for a debugger to attach before executing code."
+
+      property :watch_paths, Array,
+               description: "An array of paths which, if any are modified, will cause a job to be started."
+
+      property :working_directory, String,
+               description: "chdir to this directory, and then run the job."
```
Comment (coderanger) [inline lib/chef/resource/launchd.rb]: Missing initial capital letter.

File: lib/chef/resource/log.rb [hunk 9]
```diff
@@ -37,8 +37,13 @@ class Log < Chef::Resource
                   " entry that is not built into the resource collection, use Chef::Log instead"\
                   " of the log resource.)"
 
-      property :message, String, name_property: true, identity: true
-      property :level, Symbol, equal_to: [ :debug, :info, :warn, :error, :fatal ], default: :info
+      property :message, String,
+               name_property: true, identity: true,
+               description: "The message to be added to a log file. If not specified we'll use the resource's name instead."
```
Comment (coderanger) [inline lib/chef/resource/log.rb]: We should probably come up with a standard for how to document name_props re: the "If not specified ..." bit. But this is fine for now certainly :)

File: lib/chef/resource/log.rb [hunk 10]
```diff
@@ -37,8 +37,13 @@ class Log < Chef::Resource
...
+
+      property :level, Symbol,
+               equal_to: [ :debug, :info, :warn, :error, :fatal ], default: :info,
+               description: "The level of logging that is to be displayed by the Chef Client."
```
Comment (coderanger) [inline lib/chef/resource/log.rb]: This makes it sound like its setting `Chef::Log.level`. Maybe "The logging level to display this message at."?

File: lib/chef/resource/mdadm.rb [hunk 11]
```diff
@@ -32,14 +32,35 @@ class Mdadm < Chef::Resource
...
-      property :bitmap, String
-      property :raid_device, String, identity: true, name_property: true
-      property :layout, String
+      property :chunk, Integer,
+               default: 16,
+               description: "The chunk size. This property should not be used for a RAID 1 mirrored pair (i.e. when the level property is set to 1)."
+
+      property :devices, Array,
+               default: lazy { [] },
+               description: "A comma-separated list of devices to be part of a RAID array"
```
Comment (coderanger) [inline lib/chef/resource/mdadm.rb]: This makes it sound like it takes a string, rather than an array.

File: lib/chef/resource/mdadm.rb [hunk 12]
```diff
@@ -32,14 +32,35 @@ class Mdadm < Chef::Resource
...
+               default: false,
+               description: "Indicates whether the RAID array exists."
+
+      property :level, Integer,
+               default: 1,
+               description: "The RAID level."
+
+      property :metadata, String,
+               default: "0.90",
+               description: "The superblock type for RAID metadata"
```
Comment (coderanger) [inline lib/chef/resource/mdadm.rb]: Missing trailing period.

**Comment by tas50** [comment]
So the wording is entirely copy / paste from the docs site except for some basic name property explanation. So it came over warts and all.

**Pull request merged**